assets/js/components/Dashboard/NewPaMessage/StationsAndZones/SelectZonesPage.tsx

```diff
@@ -0,0 +1,541 @@
+  value: string[];
+  onChange: (signIds: string[]) => void;
+  onSubmit: (signIds: string[]) => void;
+  navigateTo: (page: Page) => void;
+  places: Place[];
+}
+
+const LEFT_ZONES = ["s", "w"] as const;
+const MIDDLE_ZONES = ["c", "m"] as const;
+const RIGHT_ZONES = ["n", "e"] as const;
+
+const SelectZonesPage = ({
+  value,
+  onChange,
+  onSubmit,
+  navigateTo,
+  places,
+}: Props) => {
+  const [selectedRouteFilter, setSelectedRouteFilter] = useState("");
+  const [newPlaces, setNewPlaces] = useState<Place[]>([]);
+  const [selectedRoutes, setSelectedRoutes] = useState<Dictionary<string[]>>(
+    {},
+  );
+  const routeToRouteIDMap = useRouteToRouteIDsMap();
+
+  const isSelected = (id: string) => value.includes(id);
+
+  const directionLabels = useMemo(() => {
+    switch (selectedRouteFilter) {
+      case "Green":
+      case "Green-B":
+      case "Green-C":
+      case "Green-D":
+      case "Green-E":
+      case "Blue":
+        return {
+          left: "Westbound",
+          middle: "Westbound & Eastbound",
+          right: "Eastbound",
+        };
+      case "Silver":
+        return { left: "To South Station/Downtown", right: "Eastbound" };
+      case "Bus":
+        return { left: "", middle: "", right: "" };
+      default:
+        return {
+          left: "Southbound",
+          middle: "Southbound & Northbound",
+          right: "Northbound",
+        };
+    }
+  }, [selectedRouteFilter]);
+
+  const massSelectButtonLabels = useMemo(() => {
+    if (selectedRouteFilter === "Silver") {
+      return {
+        left: "Both",
+        middle: "South Station/Downtown",
+        right: "Outbound Content",
+      };
+    }
+
+    return {
+      left: "All Zones (except bus)",
+      middle: directionLabels.left,
+      right: directionLabels.right,
+    };
+  }, [selectedRouteFilter]);
+
+  useEffect(() => {
+    const routes = fp.flow(
+      fp.flatMap((place: Place) =>
+        place.screens.filter((s) => value.includes(s.id)),
+      ),
+      fp.flatMap((screen: Screen) => screen.route_ids),
+      fp.uniq,
+      fp.groupBy((routeID: string) => {
+        if (routeID.startsWith("Green")) {
+          return "Green";
+        }
+
+        if (routeToRouteIDMap["Bus"].includes(routeID)) {
+          return "Bus";
+        }
+
+        if (routeToRouteIDMap["Silver"].includes(routeID)) {
+          return "Silver";
+        }
+
+        return routeID;
+      }),
+    )(places);
+
+    setSelectedRoutes(routes);
+    setSelectedRouteFilter(Object.keys(routes)[0]);
+    setNewPlaces(
+      places.filter((p) =>
+        places
+          .filter((p) => p.screens.some((s) => value.includes(s.id)))
+          .map((p) => p.id)
+          .includes(p.id),
+      ),
+    );
+  }, [places]);
+
+  const filteredPlaces = useMemo(() => {
+    return newPlaces.filter((p) =>
+      p.screens.some((s) =>
+        s.route_ids?.some((r) =>
+          routeToRouteIDMap[selectedRouteFilter]?.some((a) => r === a),
```

Comment:
**note:** i feel like i am seeing this pattern _a lot_. might be a good opportunity for a helper function.